packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/components/Task/Task.js

```diff
@@ -0,0 +1,46 @@
+import { Box, Paragraph } from 'grommet'
+import { bool, func, shape } from 'prop-types'
+import React, { useEffect, useState } from 'react'
+
+import taskRegistry from '@plugins/tasks'
+
+export default function Task (props) {
+  const { classification, disabled, task } = props
+  const [ annotation, setAnnotation ] = useState()
+  const { TaskComponent } = taskRegistry.get(task.type)
+
+  useEffect(function onMount() {
+    const annotation = classification.addAnnotation(task)
+    task.setAnnotation(annotation)
+    setAnnotation(annotation)
+  }, [])
```

Comment:
This makes me nervous to have the annotation stored in the stores and in local component state. It seems risky as there can be potential of getting out of sync with each other. From testing, this works, but I really would like us to rethink how annotations are being stored with an aim of simplification.

Reply:
I've got another commit, which I've pushed up as #1870, which replaces `useEffect` and `useState` with a `useStores` hook, so the `Task` component becomes a generic container which takes a `task` prop, connects to the store and passes `annotation` and `disabled` props down to a specific task.

I haven't tested that branch in depth, but the disadvantage of that might be that the `addAnnotation` action gets called more than once during rendering.

Reply:
#1870 doesn’t fix the issue this PR is addressing (actions being called during render.) It refractors the actions into a function which is called on every render, but `useStores` isn’t a hook. After reading the [docs for custom hooks](https://reactjs.org/docs/hooks-custom.html), I think `useStores` still needs to use `useEffect` and `useState`.

Reply:
`useStores` here should now be a hook which only runs `classification.addAnnotation(task)` on component mount.

Reply:
What we'd really like, I think, is to add/create an annotation on mount (if one doesn't already exist), then access `classification.annotation(task)`. `classification.annotation(task)` is a pure view now, so it shouldn't matter how many times we call it (I think.)

Reply:
You can [set references from `useEffect`](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables), so I've experimented with using `useRef` instead of `useState`. However, that breaks I Fancy Cats. I think that's happening because changing a ref doesn't result in a new render.

Reply:
I don't understand why you changed again what was in #1870. This now still has the same issue that I had concerns about before, which was the fact we're storing the annotation in component state and in the MST state. I don't think we need a custom hook or local component state here. It functioned as it was proposed and accepted in #1870 with `useContext` which triggers the rerender as needed: https://reactjs.org/docs/hooks-reference.html#usecontext

Reply:
Without the effect hook, `classification.addAnnotation(task)` is called on every render, which leads to the bug where the store throws errors like this:

```
Uncaught Error: [mobx-state-tree] You are trying to read or write to an object that is 
no longer part of a state tree. 
(Object type: 'TextAnnotation', 
Path upon death: '/classifications/resources/ckgullo6r00002a6anqcvzwmy/annotations/ckgullodx00012a6ae99dqhcj', 
Subpath: 'id', Action: ''). 
Either detach nodes first, or don't use objects after removing / replacing them in the tree.
```

I'm trying to mitigate that here by making sure that `addAnnotation` is only called once for a task.

Reply:
There might be a simpler solution, now that `classification.annotation(task)` doesn't create new annotations. :thinking:

Reply:
I think I'm happier with the code now. It checks the `classification.annotation(task)` on every context change, but only runs `classification.addAnnotation(task)` if there's no annotation already.

Reply:
I didn't see those errors in #1870 while testing, but conceptually it makes sense that it could happen. I see that you've updated the branch to remove the local state, so thanks for doing that. I can understand the desire to get `useContext` to run once since that seems like what we want, however I think the mix of `useContext` and `useState`/`useEffect` for the same state seems potentially error prone. `useContext` is intended to rerun if the context changes and the way the `useStores` function is written, it should rerun if either the store or task prop changes. I like the explicit check for the existence of `annotation` now and makes it more clear to me what the intention is: only call `addAnnotation` if `annotation` doesn't exist. I think what we have now is preferred.

FWIW, it's my understanding that the empty dependency array of useEffect is to be avoided... it's often what people have recommended as the `componentDidMount` "replacement" but apparently that can be error prone. I've found these two articles helpful to understand this:

https://reacttraining.com/blog/useEffect-is-not-the-new-componentDidMount/
https://medium.com/better-programming/understanding-the-useeffect-dependency-array-2913da504c44

I find it helpful to remember the array is our way of telling React what the effect in useEffect is dependent on and an empty array should only be used when the effect truly _depends on nothing_. So suppose we had a scenario where useContext wasn't being used, but we wanted something to run once, I _think_ the preferred method is this over an empty dependency array:
```js
const [state, setComponentState] = useState(null)
useEffect(function myEffect () {
  const newState = doSomething()
  setComponentState(newState)
}, [setComponentState])
```
So I think it will only run once upon mount and clearly declares the function as a dependency in the effect. We haven't really been doing this, but the second article helped clarify for me that you can actually put functions in the dependency array. (I definitely could still be wrong about this, as I'm still wrapping my brain around it. It was easy to just grab the simple solution of using empty arrays in previous solutions with `useEffect`, but I'm going to try to avoid it going forward).

Reply:
Thanks! I didn’t know that about `useEffect` either. I will give those articles a read tomorrow. 


I think the acid test for this PR would be: does it fix the focus bug that stops you from typing into text tasks? IIRC that was being caused by the observers.

Reply:
I'm not seeing the focus bug, but would be good to maybe try out each other's test projects. Which project do you use to test the text task?

Reply:
Mark mentioned it in a PR, but I forget which one. It shows up if you have a text task in the main task area, not in a popup. Every character you type, the task component unmounts and mounts again, which stops you from typing more than one letter.

Reply:
https://github.com/zooniverse/front-end-monorepo/pull/1868#issuecomment-716817616

Reply:
Confirmed that there is no focus bug with https://localhost:8080/?project=1693&workflow=3430

Reply:
By the way, referential integrity, from that second post, is the reason to avoid binding handlers with `onClick={handleClick.bind(this)}` in components. Every time the component renders, a new copy of `onClick` (new reference) is created.

Reply:
The closure in the first post reminded me of this example (updated to use JSX.) I remember asking what this code does being a JS interview question, before `let` existed:
```jsx
const items = []
for (var i = 1; i < 4; i++) {
  items.push(<button onClick={() => alert(i)}>{i}</button>)
}
return(<>{items}</>)
```